src/azul/service/avro_pfb.py

```diff
@@ -167,24 +167,37 @@ def _add_missing_fields(cls, name: str, object_: MutableJSON, schema):
         None is the default value, but because of https://github.com/DataBiosphere/azul/issues/2370
         this isn't currently reflected in the schema.
         """
-        object_schema = one(f for f in schema['fields'] if f['name'] == 'object')
-        entity_schema = one(e for e in object_schema['type'] if e['name'] == name)
+        if schema['type'] == 'record':
+            object_schema = one(f for f in schema['fields'] if f['name'] == 'object')
+            entity_schema = one(e for e in object_schema['type'] if e['name'] == name)
+        elif isinstance(schema['type'], dict):
+            entity_schema = schema['type']['items']
+        else:
+            assert False, schema
         for field in entity_schema['fields']:
             if field['name'] not in object_:
-                if name == 'files':
+                if isinstance(field['type'], list):
+                    # FIXME: Change 'string' to 'null'
+                    #        https://github.com/DataBiosphere/azul/issues/2462
+                    assert 'string' in field['type'] or 'null' in field['type'], field
                     default_value = None
-                else:
-                    assert field['type']['type'] == 'array'
-                    # Default for a list of records is an empty list
+                elif field['type']['type'] == 'array':
                     if isinstance(field['type']['items'], dict):
-                        assert field['type']['items']['type'] == 'record'
+                        assert field['type']['items']['type'] == 'record', field
                         default_value = []
-                    # All other types are lists of primitives where None is an accepted value
                     else:
-                        # FIXME: Change 'string' to 'null' in https://github.com/DataBiosphere/azul/issues/2462
-                        assert 'string' in field['type']['items']
+                        # FIXME: Change 'string' to 'null'
+                        #        https://github.com/DataBiosphere/azul/issues/2462
+                        assert 'string' in field['type']['items'], field
                         default_value = [None]
+                else:
+                    assert False, field
                 object_[field['name']] = default_value
```

Comment:
It seem that there is another semantically neutral refactoring in this hunk that could be a separate commit.